dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -30,79 +30,48 @@ void checkPacketSize(size_t size)
         throw Exception(fmt::format("Packet is too large to send, size : {}", size));
 }
 
+TrackedMppDataPacketPtr serializePacket(tipb::SelectResponse & response)
+{
+    auto tracked_packet = std::make_shared<TrackedMppDataPacket>();
+    tracked_packet->serializeByResponse(response);
+    checkPacketSize(tracked_packet->getPacket().ByteSizeLong());
+    return tracked_packet;
+}
 } // namespace
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::clearExecutionSummaries(tipb::SelectResponse & response)
+void MPPTunnelSetBase<Tunnel>::sendExecutionSummary(tipb::SelectResponse & response)
 {
-    /// can not use response.clear_execution_summaries() because
-    /// TiDB assume all the executor should return execution summary
-    for (int i = 0; i < response.execution_summaries_size(); i++)
-    {
-        auto * mutable_execution_summary = response.mutable_execution_summaries(i);
-        mutable_execution_summary->set_num_produced_rows(0);
-        mutable_execution_summary->set_num_iterations(0);
-        mutable_execution_summary->set_concurrency(0);
-    }
+    RUNTIME_CHECK(!tunnels.empty());
+    // for execution summary, only need to send to one tunnel.
+    tunnels.back()->write(serializePacket(response));
 }
 
 template <typename Tunnel>
 void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
 {
-    TrackedMppDataPacket tracked_packet;
-    tracked_packet.serializeByResponse(response);
-    tunnels[0]->write(tracked_packet.getPacket());
-
-    if (tunnels.size() > 1)
-    {
-        /// only the last response has execution_summaries
-        if (response.execution_summaries_size() > 0)
-        {
-            clearExecutionSummaries(response);
-            tracked_packet = TrackedMppDataPacket();
-            tracked_packet.serializeByResponse(response);
-        }
-        for (size_t i = 1; i < tunnels.size(); ++i)
-            tunnels[i]->write(tracked_packet.getPacket());
-    }
+    // for root mpp task, only one tunnel will connect to tidb/tispark.
+    RUNTIME_CHECK(1 == tunnels.size());
+    tunnels.back()->write(serializePacket(response));
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::write(mpp::MPPDataPacket & packet)
+void MPPTunnelSetBase<Tunnel>::broadcastWrite(const TrackedMppDataPacketPtr & packet)
 {
-    checkPacketSize(packet.ByteSizeLong());
+    checkPacketSize(packet->getPacket().ByteSizeLong());
+    RUNTIME_ASSERT(!tunnels.empty());
+    //  TODO avoid copy packet for broadcast.
```

Comment:
Why leave this TODO here? AFAIK, there is no risk to share the packet between tunnels, the only special case is local tunnel(due to memory tracker switch), I think we can only copy the packet for local tunnel.

Reply:
Because I was worried that the switch memory tracker in MPPTunnel would introduce problems.

Reply:
I think just copy for local tunnel is ok.

Reply:
ok

Reply:
It would be nice to hide this detail in tunnel itself. Just a suggestion.

Reply:
I agree, but it's hard to do at the moment :(
Because the copy packet is needed when and only when `tunnels.size() > 1 && is_local`